WooCommerce/Classes/ViewRelated/Products/Edit Product/Downloadable Files/Add or Edit FIle/ProductDownloadFileViewModel.swift

```diff
@@ -0,0 +1,184 @@
+
+    // Navigation actions
+    func completeUpdating(onCompletion: ProductDownloadFileViewController.Completion, onError: (ProductDownloadFileError) -> Void)
+    func hasUnsavedChanges() -> Bool
+}
+
+/// Error cases that could occur in product download file settings.
+///
+enum ProductDownloadFileError: Error {
+    case emptyFileName
+    case emptyFileUrl
+    case invalidFileUrl
+}
+
+/// Provides view data for downloadable file, and handles init/UI/navigation actions needed in product downloadable file settings.
+///
+final class ProductDownloadFileViewModel: ProductDownloadFileViewModelOutput {
+    private let product: ProductFormDataModel
+    private let downloadableFileIndex: Int
+
+    // Editable data
+    //
+    private(set) var fileName: String?
+    private(set) var fileURL: String?
+    private(set) var fileID: String?
+    private(set) var formType: ProductDownloadFileViewController.FormType
+
+    // Validation
+    private var fileNameIsValid: Bool = false
+    private var fileUrlIsValid: Bool = false
+    private lazy var throttler: Throttler = Throttler(seconds: 0.5)
+
+    init(product: ProductFormDataModel,
+         downloadFileIndex: Int = -1,
+         formType: ProductDownloadFileViewController.FormType) {
+        self.product = product
+        self.downloadableFileIndex = downloadFileIndex
+        self.formType = formType
+
+        if downloadFileIndex >= 0 {
+            let file = product.downloadableFiles[downloadFileIndex]
+            fileName = file.name
+            fileURL = file.fileURL
+            fileID = file.downloadID
+        }
+    }
+
+    var sections: [Section] {
+        let nameSection = Section(footer: Strings.fileNameFooter, rows: [.name])
+        let urlSection = Section(footer: Strings.urlFooter, rows: [.url])
+
+        switch product {
+        case is EditableProductModel:
+            return [
+                urlSection,
+                nameSection
+            ]
+        default:
+            fatalError("Unsupported product type: \(product)")
+        }
+    }
+}
+
+// MARK: - UI changes
+//
+extension ProductDownloadFileViewModel: ProductDownloadFileActionHandler {
+    func handleFileNameChange(_ fileName: String?, onValidation: @escaping (_ isValid: Bool, _ shouldBringUpKeyboard: Bool) -> Void) {
+        self.fileName = fileName
+
+        let newValue = self.fileName
+        var oldValue: String?
+        if downloadableFileIndex >= 0 {
+            oldValue = product.downloadableFiles[downloadableFileIndex].name
+        }
+
+        guard newValue != oldValue else {
+            fileNameIsValid = false
+            onValidation(fileNameIsValid || fileUrlIsValid, true)
+            return
+        }
+
+        if newValue?.isEmpty == false {
+            fileNameIsValid = true
+        } else {
+            fileNameIsValid = false
+        }
+
+        onValidation(isChangesValid(), false)
+    }
```

Comment:
Nit on code improvements:
```suggestion
        self.fileName = fileName

        let newFileName = fileName
        let oldFileName = product.downloadableFiles[safe: downloadableFileIndex]?.name

        guard newFileName != oldFileName && newFileName?.isEmpty == false else {
            fileNameIsValid = false
            onValidation(fileNameIsValid || fileUrlIsValid, true)
            return
        }

        fileNameIsValid = true

        onValidation(isChangesValid(), false)
    }
```

Reply:
updated